source/opt/split_combined_image_sampler_pass.cpp

```diff
@@ -0,0 +1,611 @@
+      default: {
+        auto* combined_ty = def_use_mgr_->GetDef(combined->type_id());
+        if (combined_ty->opcode() == spv::Op::OpTypeSampledImage) {
+          // This operates on the sampled image. We're removing the sampled
+          // image: recreate it.
+          // Example: OpImage, OpImageSampleExplicitLod, etc.
+          builder.SetInsertPoint(use.user);
+          auto* sampled_image = builder.AddSampledImage(
+              combined_ty->result_id(), use.image_part->result_id(),
+              use.sampler_part->result_id());
+          use.user->SetOperand(use.index, {sampled_image->result_id()});
+          def_use_mgr_->AnalyzeInstUse(use.user);
+          break;
+        }
+        return Fail() << "unhandled user: " << *use.user;
+      }
+    }
+  }
+
+  for (auto* inst : reanalyze_set) {
+    def_use_mgr_->AnalyzeInstUse(inst);
+  }
+
+  return SPV_SUCCESS;
+}
+
+spv_result_t SplitCombinedImageSamplerPass::RemapFunctions() {
+  // Remap function types. A combined type can appear as a parameter, but not as
+  // the return type.
+  {
+    std::unordered_set<Instruction*> reanalyze_set;
+    for (auto& inst : context()->types_values()) {
+      if (inst.opcode() == spv::Op::OpTypeFunction) {
+        analysis::Function* f_ty =
+            type_mgr_->GetType(inst.result_id())->AsFunction();
+        std::vector<const analysis::Type*> new_params;
+        for (const auto* param_ty : f_ty->param_types()) {
+          const auto param_ty_id = type_mgr_->GetId(param_ty);
+          if (combined_types_.find(param_ty_id) != combined_types_.end()) {
+            auto* param_type = def_use_mgr_->GetDef(param_ty_id);
+            auto [image_type, sampler_type] = SplitType(*param_type);
+            assert(image_type);
+            assert(sampler_type);
+            // The image and sampler types must already exist, so there is no
+            // need to move them to the right spot.
+            new_params.push_back(type_mgr_->GetType(image_type->result_id()));
+            new_params.push_back(type_mgr_->GetType(sampler_type->result_id()));
+          } else {
+            new_params.push_back(param_ty);
+          }
+        }
+        if (new_params.size() != f_ty->param_types().size()) {
+          // Replace this type.
+          analysis::Function new_f_ty(f_ty->return_type(), new_params);
+          const uint32_t new_f_ty_id = type_mgr_->GetTypeInstruction(&new_f_ty);
+          def_use_mgr_->ForEachUse(&inst,
+                                   [&](Instruction* user, uint32_t use_index) {
+                                     user->SetOperand(use_index, {new_f_ty_id});
+                                     reanalyze_set.insert(user);
+                                   });
+          MarkAsDead(&inst);
+
+          reanalyze_set.insert(def_use_mgr_->GetDef(new_f_ty_id));
+          // Reanalyze the non-combined parameter types, and the return type.
+          auto* new_f_ty_inst = def_use_mgr_->GetDef(new_f_ty_id);
```

Comment:
```suggestion
          // Reanalyze the non-combined parameter types, and the return type.
          auto* new_f_ty_inst = def_use_mgr_->GetDef(new_f_ty_id);
          reanalyze_set.insert(new_f_ty_inst);
```

Reply:
The reanalysis was buggy.  Replaced this code.